pkg/custom-plugins/spec.go

```diff
@@ -0,0 +1,131 @@
+package customplugins
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/yaml"
+)
+
+// Specs is a list of plugin specs.
+type Specs []Spec
+
+// Validate validates all the plugin specs.
+func (specs Specs) Validate() error {
+	all := make(map[string]struct{})
+	for i := range specs {
+		if err := specs[i].Validate(); err != nil {
+			return err
+		}
+
+		if _, ok := all[specs[i].ComponentName()]; ok {
+			return fmt.Errorf("duplicate component name: %s", specs[i].ComponentName())
+		}
+		all[specs[i].ComponentName()] = struct{}{}
+	}
+
+	return nil
+}
+
+// LoadSpecs loads the plugin specs from the given path.
+func LoadSpecs(path string) (Specs, error) {
+	yamlFile, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var pluginSpecs Specs
+	if err := yaml.Unmarshal(yamlFile, &pluginSpecs); err != nil {
+		return nil, err
+	}
+
+	for i := range pluginSpecs {
+		if err := pluginSpecs[i].Validate(); err != nil {
+			return nil, err
+		}
+	}
+
+	return pluginSpecs, nil
+}
+
+// Spec is a plugin spec and configuration.
+// Each spec represents a single state or event, in the external-plugin component.
+type Spec struct {
```

Comment:
shall we move it into types.go and reorder the type and functions to keep a consistent view to help understand the relationship between them?